dulwich/porcelain.py

```diff
@@ -2100,113 +2101,167 @@ def checkout_branch(repo, target: Union[bytes, str], force: bool = False) -> Non
+
+    # --- 1) Read or write patterns ---
+    if patterns is None:
+        lines = repo.get_sparse_checkout_patterns()
+        if lines is None:
+            raise Error("No sparse checkout patterns found.")
+    else:
+        lines = patterns
+        sp_path = os.path.join(repo.path, ".git", "info", "sparse-checkout")
+        ensure_dir_exists(os.path.dirname(sp_path))
+        with open(sp_path, "w") as f:
+            for p in patterns:
+                f.write(p + "\n")
+        repo.set_sparse_checkout_patterns(patterns)
+
+    # --- 2) Determine the set of included paths ---
+    included_paths = determine_included_paths(repo, lines, cone)
+
+    # --- 3) Apply those results to the index & working tree ---
+    apply_included_paths(repo, included_paths, force=force)
+
+
+def cone_mode_init(repo):
+    """Initialize a repository to use sparse checkout in 'cone' mode.
+
+    Sets ``core.sparseCheckout`` and ``core.sparseCheckoutCone`` in the config.
+    Writes an initial ``.git/info/sparse-checkout`` file that includes only
+    top-level files (and excludes all subdirectories), e.g. ``["/*", "!/*/"]``.
+    Then performs a sparse checkout to update the working tree accordingly.
+
+    If no directories are specified, then only top-level files are included:
+    https://git-scm.com/docs/git-sparse-checkout#_internalscone_mode_handling
+
+    Args:
+      repo: Path to the repository or a Repo object.
+
+    Returns:
+      None
+    """
+    if not isinstance(repo, Repo):
+        repo = Repo(repo)
+
+    config = repo.get_config()
+    config.set((b"core",), b"sparseCheckout", b"true")
+    config.set((b"core",), b"sparseCheckoutCone", b"true")
+    config.write_to_path()
+
+    patterns = ["/*", "!/*/"]  # root-level files only
+    sparse_checkout(repo, patterns, force=True, cone=True)
+
+
+def cone_mode_set(repo, dirs, force=False):
+    """Overwrite the existing 'cone-mode' sparse patterns with a new set of directories.
+
+    Ensures ``core.sparseCheckout`` and ``core.sparseCheckoutCone`` are enabled.
+    Writes new patterns so that only the specified directories (and top-level files)
+    remain in the working tree, and applies the sparse checkout update.
+
+    Args:
+      repo: Path to the repository or a Repo object.
+      dirs: List of directory names to include.
+      force: Whether to forcibly discard local modifications (default False).
+
+    Returns:
+      None
+    """
+    if not isinstance(repo, Repo):
+        repo = Repo(repo)
+
+    config = repo.get_config()
+    config.set((b"core",), b"sparseCheckout", b"true")
+    config.set((b"core",), b"sparseCheckoutCone", b"true")
+    config.write_to_path()
+
+    # Initial lines: include top-level files, then exclude all subdirectories
+    new_patterns = ["/*", "!/*/"]
+
+    # For each directory to include, add an exclusion-style line that "undoes"
+    # the prior 'exclude' and re-includes that directory (and everything under it).
+    for d in dirs:
```

Comment:
Move this code to a separate function in .sparse_patterns?

Reply:
Yep, there was a lot of refactoring to be done for this to be porcelain tier code, thanks for the nudge, I've made dirs a porcelain-module concept and patterns are prepared/retrieved in the repo method.